### ISSUE TITLE
[New rule] check-for-equals-in-render-object-setters when implementing a RenderObject; maybe even apply this linter to check potential bugs in Flutter source code

### ISSUE THREAD

**Issue by fzyzcjy** [dart-code-checker/dart-code-metrics#997]
**Please describe what the rule should do:**

As long as we write down a few `RenderObject`s, or read through some in flutter source code, we will realize a pattern:

```
  set a(A a) {
    if (_a == a) return; // <---- NOTE
    _a = a;
    markNeedsPaint(); // or something else etc
  }
```

That early return is very helpful, because `set a` is always called within updateRenderObject. If it does not early return, then the markNeedsPaint (or for other fields, maybe markNeedsLayout, some heavy work, etc), will be *unconditionally* executed every time the render object is updated, even if that is totally unnecessary. 

Thus, this rule should find out such problem and lint it.

Btw, maybe we can even apply this linter to check potential bugs in Flutter source code.

**If your rule is inspired by other please provide link to it:**

**What category of rule is this? (place an "X" next to just one item)**

- [x] Warns about a potential error (problem)
- [ ] Suggests an alternate way of doing something (suggestion)
- [ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about (it will be better if you can provide both good and bad examples):**

```dart
class A extends RenderBox {
  set a(A a) {
    if (_a == a) return; // <---- if no such line, should LINT
    _a = a;
    markNeedsPaint(); // or something else etc
  }
}
```

**Are you willing to submit a pull request to implement this rule?**

**Comment by incendial** [comment]
@fzyzcjy could you share an example from Flutter codebase?

**Comment by incendial** [comment]
I look at this one https://github.com/flutter/flutter/blob/5bc4a7612f7fd430933b046aeb9e36a5c7f3916e/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L667 and it looks too hard to extract a formal rule coz of `_backButton = _updateChild` calls.

**Comment by incendial** [comment]
The info I'm looking for: which classes should be checked and which setters should have the check.

**Comment by incendial** [comment]
Right now it looks like it's `a class, that extends RenderObject` and `all setters`, but the example above breaks this rule

**Comment by fzyzcjy** [comment]
@incendial 

> could you share an example from Flutter codebase?

Again, any RenderBox's subclass. For example, RenderPadding, RenderOpacity, RenderAnythingYouLike.

**Comment by incendial** [comment]
Is it okay if I skip the check for a setter if the type is `RenderBox`?

**Comment by incendial** [comment]
Kinda looks like a pattern, but I'm not sure

<img width="1065" alt="Screenshot 2022-09-16 at 14 40 17" src="https://user-images.githubusercontent.com/21116742/190621254-f9f8110f-81cc-45aa-8f02-6c9e38e6d92b.png">

**Comment by fzyzcjy** [comment]
> 
I look at this one https://github.com/flutter/flutter/blob/5bc4a7612f7fd430933b046aeb9e36a5c7f3916e/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L667 and it looks too hard to extract a formal rule coz of _backButton = _updateChild calls.

There are outliers, surely. I guess we only need to focus on the standard things, and `// ignore: the-rule-name` outliers. For example, this is standard:

https://github.com/flutter/flutter/blob/5bc4a7612f7fd430933b046aeb9e36a5c7f3916e/packages/flutter/lib/src/cupertino/text_selection_toolbar.dart#L698-L701

**Comment by fzyzcjy** [comment]
> Is it okay if I skip the check for a setter if the type is RenderBox?

I never realized this pattern.

IMHO, maybe we can firstly implement a naive version. Then let the version report a lot of false positives. Then we check them, and see what common pattern they have. Then change our code to ignore them.

**Comment by incendial** [comment]
> IMHO, maybe we can firstly implement a naive version. Then let the version report a lot of false positives. Then we check them, and see what common pattern they have. Then change our code to ignore them.

Agreed!

**Comment by incendial** [comment]
Available in 4.19.0 🚀

**Comment by fzyzcjy** [comment]
🚀
